parlai/core/teachers.py

```diff
@@ -1944,6 +1945,223 @@ def shutdown(self):
             t.shutdown()
 
 
+class ChunkTeacher(FixedDialogTeacher, ABC):
+    """
+    Useful for loading large amounts of data.
+
+    Data is separated into chunks and loaded one chunk at a
+    time. Loads the data off of the main thread.
+    """
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+
+        # make logging print to stdout
+        logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
+
+        self.buffersize = 100000
+
+        if 'stream' not in opt['datatype']:
+            raise ValueError(
+                'Chunk teacher should be used with streaming. '
+            )
+        if opt.get('pytorch_teacher_task') is not None:
+            raise ValueError(
+                'Chunk teacher is not compatible with pytorch '
+                'data teacher.'
+            )
+        if opt['numthreads'] > 1:
+            raise ValueError(
+                'Chunk teacher is not compatible with Hogwild.'
+            )
+
+        self.set_datasettings(opt['datatype'])
+        chunks = self.fold_chunks[:]
+
+        if (
+            shared is None
+            and self.is_train
+            and self.opt.get('distributed_world_size') is not None
+        ):
+            dws = int(self.opt['distributed_world_size'])
+            rank = int(self.opt['rank'])
+            self.fold_chunks = [c for c in self.fold_chunks if c % dws == rank]
```

Comment:
Nothing you need to change here, but this will need to be carefully revisited in some of the aspects of #2185.